src/rbdmap-generator

```diff
@@ -0,0 +1,273 @@
+#!/usr/bin/python3
+
+import os
+import re
+import string
+import sys
+import logging
+import syslog
+from typing import Optional
+
+
+logger = logging.getLogger('rbdmap-generator')
+
+RBDMAP_PATH = os.environ.get('RBDMAPFILE', '/etc/ceph/rbdmap')
+
+RBDMAP_TEMPLATE = '''# Automatically generated by rbdmap-generator
+
+[Unit]
+SourcePath={rbdmap_path}
+Description=Map RBD device %I
+Documentation=man:rbdmap-generator(8)
+DefaultDependencies=no
+IgnoreOnIsolate=true
+Conflicts=umount.target
+Wants=blockdev@dev-rbd-%i.target network-online.target
+Before=blockdev@dev-rbd-%i.target
+After=remote-fs-pre.target rbdmap-pre.target systemd-udevd-kernel.socket network.target network-online.target
+{unit_extra}
+
+[Service]
+Type=oneshot
+RemainAfterExit=yes
+ExecStart=/usr/bin/rbd device map %I{rdb_map_args}
+ExecStartPost=-/etc/ceph/rbd.d/%I map /dev/rbd/%I
+ExecStop=-/etc/ceph/rbd.d/%I unmap /dev/rbd/%I
+ExecStop=/usr/bin/rbd device unmap %I{rbd_unmap_args}
+'''
+
+VALID_CHARS = string.ascii_letters + string.digits + ':_.'
+
+def systemd_escape(p: str, path=True):
+    res = ''
+    for c in p:
+        if c == '/':
+            res += '-'
+        elif c in VALID_CHARS:
+            res += c
+        else:
+            res += ''.join(R'\x{:02x}'.format(b) for b in c.encode())
+    if res[0] == '.':
+        res = R'\x2e' + res[1:]
+    return res
+
+def systemd_escape_path(p: str):
+    if (p == '/'):
+        return '-'
+    p = p.strip('/')
+    p = re.sub(r'/{2,}', '/', p)
+    return systemd_escape(p)
+
+def normalize_rbd_spec(spec: str):
+    if spec.startswith('/') or spec.endswith('/'):
+        raise ValueError(
+            'rbd image or snapshot specification {} cannot starts or ends with /'
+            .format(spec))
+    if '/' not in spec:
+        spec = 'rbd/' + spec
```

Comment:
I agree.  Let's be explicit and make it mandatory.

Reply:
This is not easy. The default pool is 'rbd', this is already documented [here](https://docs.ceph.com/en/latest/man/8/rbd/#image-snap-group-and-journal-specs), so I think this is reliable. And we cannot only let the `rbd device map` command determine the default pool, because we need to know the pool name to determine the device path (`/dev/rbd/pool/image`).

So:
1. Make the pool name mandatory. But it is not mandatory for `rbd` command. They are inconsistent.
2. Rely on the already documented default pool name `rbd`.

Given the above, do you still prefer 1? I personally prefer 2 to keep things consistent.

Reply:
The default pool is not 'rbd' but what is set in `rbd_default_pool` configuration parameter. And it can be set/overridden either in global (mon centralized db) config, or in ceph.conf or in a command line argument or in `CEPH_ARGS` environmental variable.

So if you really want to support the default pool in your scripts, you would have to call e.g. `ceph config get client.$id rbd_default_pool` to get the name, instead of just using `rbd`.

Personally I would rather we didn't have this default pool at all from the start, because it makes more problems than goods. And it looks useful only for CLI to save users more typing. So let's not propagate it into the map file. I don't see any problem that pool is not mandatory for the CLI but will be mandatory for the rbdtab file.

Reply:
Ah, and `ceph config get client.$id rbd_default_pool` would not be enough. I though it would apply the local file override but it only looks the setting in the centralized db. So it is much tricker to get properly the default pool.

Reply:
OK, thanks for the explaination. I will make it mandatory. This [rbd(8)](https://github.com/ceph/ceph/blob/master/doc/man/8/rbd.rst#image-snap-group-and-journal-specs) makes me think it is just hard-coded as "rbd". Maybe we should update it.